File: ios/EggplantDetector/PlantTargeting.swift
import CoreGraphics
import CoreVideo
import Foundation

enum TargetSelectionContract {
    static let isPhase1ManualTargetingEnabled = true
    static let manualSelectionBoxSizeRatio: CGFloat = 0.28
}

struct NormalizedRect: Equatable {
    let left: CGFloat
    let top: CGFloat
    let right: CGFloat
    let bottom: CGFloat

    init(left: CGFloat, top: CGFloat, right: CGFloat, bottom: CGFloat) {
        precondition(left <= right, "left must be <= right")
        precondition(top <= bottom, "top must be <= bottom")
        self.left = left
        self.top = top
        self.right = right
        self.bottom = bottom
    }

    static func fromCenter(
        centerX: CGFloat,
        centerY: CGFloat,
        widthRatio: CGFloat,
        heightRatio: CGFloat
    ) -> NormalizedRect {
        let clampedWidth = min(max(widthRatio, 0.01), 1)
        let clampedHeight = min(max(heightRatio, 0.01), 1)
        let maxLeft = max(1 - clampedWidth, 0)
        let maxTop = max(1 - clampedHeight, 0)
        let left = min(max(centerX - clampedWidth / 2, 0), maxLeft)
        let top = min(max(centerY - clampedHeight / 2, 0), maxTop)
        return NormalizedRect(
            left: left,
            top: top,
            right: min(max(left + clampedWidth, left), 1),
            bottom: min(max(top + clampedHeight, top), 1)
        )
    }
}

struct PlantCandidate: Equatable {
    let id: String
    let box: NormalizedRect
    let confidence: Float
}

enum TargetSelectionSource: Equatable {
    case manualTap
    case tracker
}

struct SelectedPlantTarget: Equatable {
    let id: String
    let box: NormalizedRect
    let source: TargetSelectionSource
    let updatedAt: Date
}

enum TargetLossReason: Equatable {
    case reset
    case modeChanged
    case cameraInterrupted
}

enum SelectedTargetState: Equatable {
    case none
    case selected(target: SelectedPlantTarget)
    case lost(previousTarget: SelectedPlantTarget?, reason: TargetLossReason)
}

protocol PlantDetector {
    func detectCandidates(in pixelBuffer: CVPixelBuffer) -> [PlantCandidate]
}

final class PlaceholderPlantDetector: PlantDetector {
    func detectCandidates(in pixelBuffer: CVPixelBuffer) -> [PlantCandidate] {
        []
    }
}

protocol PlantTracker {
    func updateSelectedTarget(
        currentState: SelectedTargetState,
        candidates: [PlantCandidate]
    ) -> SelectedTargetState
}

final class ManualLockPlantTracker: PlantTracker {
    func updateSelectedTarget(
        currentState: SelectedTargetState,
        candidates: [PlantCandidate]
    ) -> SelectedTargetState {
        currentState
    }
}

final class SelectedTargetController {
    private let now: () -> Date

    private(set) var state: SelectedTargetState = .none

    init(now: @escaping () -> Date = Date.init) {
        self.now = now
    }

    @discardableResult
    func selectManualTarget(box: NormalizedRect) -> SelectedTargetState {
        state = .selected(
            target: SelectedPlantTarget(
                id: UUID().uuidString,
                box: box,
                source: .manualTap,
                updatedAt: now()
            )
        )
        return state
    }

    @discardableResult
    func resetTarget(reason: TargetLossReason = .reset) -> SelectedTargetState {
        switch state {
        case let .selected(target):
            state = .lost(previousTarget: target, reason: reason)
        case let .lost(previousTarget, _):
            state = .lost(previousTarget: previousTarget, reason: reason)
        case .none:
            state = .none
        }
        return state
    }

    @discardableResult
    func applyTrackedState(_ updatedState: SelectedTargetState) -> SelectedTargetState {
        state = updatedState
        return state
    }
}

enum PreviewTargetMapper {
    static func analysisSquare(in bounds: CGRect) -> CGRect {
        let size = min(bounds.width, bounds.height)
        let originX = bounds.midX - size / 2
        let originY = bounds.midY - size / 2
        return CGRect(x: originX, y: originY, width: size, height: size)
    }

    static func manualSelection(
        for point: CGPoint,
        in bounds: CGRect,
        boxSizeRatio: CGFloat = TargetSelectionContract.manualSelectionBoxSizeRatio
    ) -> NormalizedRect {
        let square = analysisSquare(in: bounds)
        let clampedX = min(max(point.x, square.minX), square.maxX)
        let clampedY = min(max(point.y, square.minY), square.maxY)
        let normalizedX = min(max((clampedX - square.minX) / square.width, 0), 1)
        let normalizedY = min(max((clampedY - square.minY) / square.height, 0), 1)
        return NormalizedRect.fromCenter(
            centerX: normalizedX,
            centerY: normalizedY,
            widthRatio: boxSizeRatio,
            heightRatio: boxSizeRatio
        )
    }

    static func viewRect(for box: NormalizedRect, in bounds: CGRect) -> CGRect {
        let square = analysisSquare(in: bounds)
        return CGRect(
            x: square.minX + box.left * square.width,
            y: square.minY + box.top * square.height,
            width: (box.right - box.left) * square.width,
            height: (box.bottom - box.top) * square.height
        )
    }
}

File: android/src/main/java/com/example/eggplantdetector/PlantTargeting.kt
package com.example.eggplantdetector

import android.graphics.Bitmap
import android.graphics.Rect
import android.graphics.RectF
import java.util.UUID
import kotlin.math.min
import kotlin.math.roundToInt

object TargetSelectionContract {
    const val isPhase1ManualTargetingEnabled = true
    const val manualSelectionBoxSizeRatio = 0.28f
}

data class NormalizedRect(
    val left: Float,
    val top: Float,
    val right: Float,
    val bottom: Float
) {
    init {
        require(left <= right) { "left must be <= right" }
        require(top <= bottom) { "top must be <= bottom" }
    }

    companion object {
        fun fromCenter(
            centerX: Float,
            centerY: Float,
            widthRatio: Float,
            heightRatio: Float
        ): NormalizedRect {
            val clampedWidth = widthRatio.coerceIn(0.01f, 1f)
            val clampedHeight = heightRatio.coerceIn(0.01f, 1f)
            val maxLeft = (1f - clampedWidth).coerceAtLeast(0f)
            val maxTop = (1f - clampedHeight).coerceAtLeast(0f)
            val left = (centerX - clampedWidth / 2f).coerceIn(0f, maxLeft)
            val top = (centerY - clampedHeight / 2f).coerceIn(0f, maxTop)
            return NormalizedRect(
                left = left,
                top = top,
                right = (left + clampedWidth).coerceIn(left, 1f),
                bottom = (top + clampedHeight).coerceIn(top, 1f)
            )
        }
    }
}

data class PlantCandidate(
    val id: String,
    val box: NormalizedRect,
    val confidence: Float
)

enum class TargetSelectionSource {
    MANUAL_TAP,
    TRACKER
}

data class SelectedPlantTarget(
    val id: String,
    val box: NormalizedRect,
    val source: TargetSelectionSource,
    val updatedAtMs: Long
)

enum class TargetLossReason {
    RESET,
    MODE_CHANGED,
    CAMERA_INTERRUPTED
}

sealed interface SelectedTargetState {
    object None : SelectedTargetState

    data class Selected(
        val target: SelectedPlantTarget
    ) : SelectedTargetState

    data class Lost(
        val previousTarget: SelectedPlantTarget?,
        val reason: TargetLossReason
    ) : SelectedTargetState
}

interface PlantDetector {
    fun detectCandidates(bitmap: Bitmap): List<PlantCandidate>
}

class PlaceholderPlantDetector : PlantDetector {
    override fun detectCandidates(bitmap: Bitmap): List<PlantCandidate> = emptyList()
}

interface PlantTracker {
    fun updateSelectedTarget(
        currentState: SelectedTargetState,
        candidates: List<PlantCandidate>
    ): SelectedTargetState
}

class ManualLockPlantTracker : PlantTracker {
    override fun updateSelectedTarget(
        currentState: SelectedTargetState,
        candidates: List<PlantCandidate>
    ): SelectedTargetState = currentState
}

class SelectedTargetController(
    private val nowMs: () -> Long = { System.currentTimeMillis() }
) {
    var state: SelectedTargetState = SelectedTargetState.None
        private set

    fun selectManualTarget(box: NormalizedRect): SelectedTargetState {
        state = SelectedTargetState.Selected(
            target = SelectedPlantTarget(
                id = UUID.randomUUID().toString(),
                box = box,
                source = TargetSelectionSource.MANUAL_TAP,
                updatedAtMs = nowMs()
            )
        )
        return state
    }

    fun resetTarget(reason: TargetLossReason = TargetLossReason.RESET): SelectedTargetState {
        state = when (val currentState = state) {
            is SelectedTargetState.Selected -> SelectedTargetState.Lost(currentState.target, reason)
            is SelectedTargetState.Lost -> SelectedTargetState.Lost(currentState.previousTarget, reason)
            SelectedTargetState.None -> SelectedTargetState.None
        }
        return state
    }

    fun applyTrackedState(updatedState: SelectedTargetState): SelectedTargetState {
        state = updatedState
        return state
    }
}

object PreviewTargetMapper {
    fun analysisSquare(viewWidth: Float, viewHeight: Float): RectF {
        val size = min(viewWidth, viewHeight)
        val left = (viewWidth - size) / 2f
        val top = (viewHeight - size) / 2f
        return RectF(left, top, left + size, top + size)
    }

    fun manualSelectionForTap(
        tapX: Float,
        tapY: Float,
        viewWidth: Float,
        viewHeight: Float,
        boxSizeRatio: Float = TargetSelectionContract.manualSelectionBoxSizeRatio
    ): NormalizedRect {
        val square = analysisSquare(viewWidth, viewHeight)
        val clampedX = tapX.coerceIn(square.left, square.right)
        val clampedY = tapY.coerceIn(square.top, square.bottom)
        val normalizedX = ((clampedX - square.left) / square.width()).coerceIn(0f, 1f)
        val normalizedY = ((clampedY - square.top) / square.height()).coerceIn(0f, 1f)
        return NormalizedRect.fromCenter(
            centerX = normalizedX,
            centerY = normalizedY,
            widthRatio = boxSizeRatio,
            heightRatio = boxSizeRatio
        )
    }

    fun viewRectFor(box: NormalizedRect, viewWidth: Float, viewHeight: Float): RectF {
        val square = analysisSquare(viewWidth, viewHeight)
        return RectF(
            square.left + (box.left * square.width()),
            square.top + (box.top * square.height()),
            square.left + (box.right * square.width()),
            square.top + (box.bottom * square.height())
        )
    }
}

object SelectedTargetCropper {
    fun analysisSquareRect(imageWidth: Int, imageHeight: Int): Rect {
        val size = min(imageWidth, imageHeight)
        val left = (imageWidth - size) / 2
        val top = (imageHeight - size) / 2
        return Rect(left, top, left + size, top + size)
    }

    fun pixelRectFor(box: NormalizedRect, imageWidth: Int, imageHeight: Int): Rect {
        val square = analysisSquareRect(imageWidth, imageHeight)
        val squareWidth = square.width().toFloat()
        val squareHeight = square.height().toFloat()
        val left = (square.left + (box.left * squareWidth)).roundToInt().coerceIn(0, imageWidth - 1)
        val top = (square.top + (box.top * squareHeight)).roundToInt().coerceIn(0, imageHeight - 1)
        val right = (square.left + (box.right * squareWidth)).roundToInt().coerceIn(left + 1, imageWidth)
        val bottom = (square.top + (box.bottom * squareHeight)).roundToInt().coerceIn(top + 1, imageHeight)
        return Rect(left, top, right, bottom)
    }

    fun cropBitmap(bitmap: Bitmap, box: NormalizedRect): Bitmap {
        val pixelRect = pixelRectFor(box, bitmap.width, bitmap.height)
        return Bitmap.createBitmap(
            bitmap,
            pixelRect.left,
            pixelRect.top,
            pixelRect.width(),
            pixelRect.height()
        )
    }
}
